File: examples/typescript/token_demo.ts
#!/usr/bin/env npx tsx
// Resource operations demonstration - TypeScript GraphQL Client
// Shows detailed resource lifecycle operations using GraphQL API
// Run with: npx tsx examples/typescript/token_demo.ts

import { CircuitBreakerClient, type ResourceGQL } from "./basic_workflow.js";

function logSuccess(message: string) {
  console.log(`✅ ${message}`);
}

function logInfo(message: string) {
  console.log(`ℹ️  ${message}`);
}

function logError(message: string) {
  console.log(`❌ ${message}`);
}

function logWarning(message: string) {
  console.log(`⚠️  ${message}`);
}

async function main() {
  console.log(
    "🚀 Circuit Breaker Resource Operations Demo - TypeScript Client",
  );
  console.log(
    "===============================================================",
  );
  console.log();

  const client = new CircuitBreakerClient();

  try {
    // Create AI content creation workflow
    logInfo("Creating AI Content Creation Workflow...");

    const workflowInput = {
      name: "AI-Powered Content Creation",
      states: [
        "ideation",
        "drafting",
        "review",
        "revision",
        "approval",
        "published",
        "archived",
      ],
      activities: [
        {
          id: "start_drafting",
          fromStates: ["ideation"],
          toState: "drafting",
          conditions: [],
        },
        {
          id: "submit_for_review",
          fromStates: ["drafting"],
          toState: "review",
          conditions: [],
        },
        {
          id: "request_revision",
          fromStates: ["review"],
          toState: "revision",
          conditions: [],
        },
        {
          id: "back_to_drafting",
          fromStates: ["revision"],
          toState: "drafting",
          conditions: [],
        },
        {
          id: "approve_content",
          fromStates: ["review"],
          toState: "approval",
          conditions: [],
        },
        {
          id: "publish_content",
          fromStates: ["approval"],
          toState: "published",
          conditions: [],
        },
        {
          id: "archive_content",
          fromStates: ["published"],
          toState: "archived",
          conditions: [],
        },
        {
          id: "back_to_ideation",
          fromStates: ["archived"],
          toState: "ideation",
          conditions: [],
        },
      ],
      initialState: "ideation",
    };

    const workflowResult = await client.createWorkflow(workflowInput);

    if (workflowResult.errors) {
      logError(
        `Failed to create workflow: ${workflowResult.errors.map((e) => e.message).join(", ")}`,
      );
      return;
    }

    const workflow = workflowResult.data!.createWorkflow;
    logSuccess(`Created workflow: ${workflow.name} (${workflow.id})`);
    console.log();

    // Create multiple content resources
    const contentTopics = [
      {
        title: "Introduction to Rust Programming",
        type: "tutorial",
        targetAudience: "beginners",
        estimatedReadTime: 15,
      },
      {
        title: "Advanced TypeScript Patterns",
        type: "guide",
        targetAudience: "intermediate",
        estimatedReadTime: 25,
      },
      {
        title: "Building Scalable APIs with GraphQL",
        type: "article",
        targetAudience: "advanced",
        estimatedReadTime: 20,
      },
    ];

    const resources: ResourceGQL[] = [];

    for (const [index, topic] of contentTopics.entries()) {
      logInfo(`Creating content resource ${index + 1}/3: ${topic.title}`);

      const resourceInput = {
        workflowId: workflow.id,
        initialState: "ideation",
        data: {
          title: topic.title,
          type: topic.type,
          targetAudience: topic.targetAudience,
          estimatedReadTime: topic.estimatedReadTime,
          keywords: topic.title.toLowerCase().split(" ").slice(0, 3),
          status: "planning",
          wordCount: 0,
          authorId: "ai-assistant",
          createdAt: new Date().toISOString(),
        },
        metadata: {
          priority: index === 1 ? "high" : "medium",
          contentType: topic.type,
          seoOptimized: false,
          socialMediaReady: false,
          version: "1.0",
        },
      };

      const resourceResult = await client.createResource(resourceInput);

      if (resourceResult.errors) {
        logError(
          `Failed to create resource: ${resourceResult.errors.map((e) => e.message).join(", ")}`,
        );
        continue;
      }

      const resource = resourceResult.data!.createResource;
      resources.push(resource);
      logSuccess(`Created resource: ${resource.data.title} (${resource.id})`);
    }

    console.log();
    logInfo(`Created ${resources.length} content resources`);

    // Demonstrate different resource lifecycles
    for (const [index, resource] of resources.entries()) {
      console.log();
      logInfo(`Processing content ${index + 1}: ${resource.data.title}`);

      let currentResource = resource;

      // Start drafting
      logInfo("Starting drafting phase...");
      const draftResult = await client.executeActivity({
        resourceId: currentResource.id,
        activityId: "start_drafting",
        data: {
          action: "start_drafting",
          timestamp: new Date().toISOString(),
          notes: "AI content generation initiated",
        },
      });

      if (draftResult.errors) {
        logError(
          `Failed to start drafting: ${draftResult.errors.map((e) => e.message).join(", ")}`,
        );
        continue;
      }

      currentResource = draftResult.data!.executeActivity;

      // Update resource data to simulate content creation
      currentResource.data.wordCount = Math.floor(Math.random() * 1000) + 500;
      currentResource.data.status = "draft_complete";

      // Submit for review
      logInfo("Submitting for review...");
      const reviewResult = await client.executeActivity({
        resourceId: currentResource.id,
        activityId: "submit_for_review",
        data: {
          action: "submit_for_review",
          timestamp: new Date().toISOString(),
          wordCount: currentResource.data.wordCount,
          readabilityScore: Math.floor(Math.random() * 40) + 60,
        },
      });

      if (reviewResult.errors) {
        logError(
          `Failed to submit for review: ${reviewResult.errors.map((e) => e.message).join(", ")}`,
        );
        continue;
      }

      currentResource = reviewResult.data!.executeActivity;

      // Simulate different review outcomes
      const reviewOutcome = index === 1 ? "revision" : "approval";

      if (reviewOutcome === "revision") {
        logWarning("Content needs revision...");

        const revisionResult = await client.executeActivity({
          resourceId: currentResource.id,
          activityId: "request_revision",
          data: {
            action: "request_revision",
            timestamp: new Date().toISOString(),
            feedback: "Needs more technical examples and clearer explanations",
          },
        });

        if (revisionResult.errors) {
          logError(
            `Failed to request revision: ${revisionResult.errors.map((e) => e.message).join(", ")}`,
          );
          continue;
        }

        currentResource = revisionResult.data!.executeActivity;

        // Back to drafting
        logInfo("Returning to drafting for revisions...");
        const backToDraftResult = await client.executeActivity({
          resourceId: currentResource.id,
          activityId: "back_to_drafting",
          data: {
            action: "back_to_drafting",
            timestamp: new Date().toISOString(),
            revisionNotes: "Incorporating reviewer feedback",
          },
        });

        if (backToDraftResult.errors) {
          logError(
            `Failed to return to drafting: ${backToDraftResult.errors.map((e) => e.message).join(", ")}`,
          );
          continue;
        }

        currentResource = backToDraftResult.data!.executeActivity;

        // Resubmit
        logInfo("Resubmitting revised content...");
        const resubmitResult = await client.executeActivity({
          resourceId: currentResource.id,
          activityId: "submit_for_review",
          data: {
            action: "resubmit_for_review",
            timestamp: new Date().toISOString(),
            revisionComplete: true,
            wordCount: currentResource.data.wordCount + 200,
          },
        });

        if (resubmitResult.errors) {
          logError(
            `Failed to resubmit: ${resubmitResult.errors.map((e) => e.message).join(", ")}`,
          );
          continue;
        }

        currentResource = resubmitResult.data!.executeActivity;
      }

      // Approve content
      logInfo("Approving content...");
      const approveResult = await client.executeActivity({
        resourceId: currentResource.id,
        activityId: "approve_content",
        data: {
          action: "approve_content",
          timestamp: new Date().toISOString(),
          approvedBy: "content-manager",
          qualityScore: Math.floor(Math.random() * 20) + 80,
        },
      });

      if (approveResult.errors) {
        logError(
          `Failed to approve content: ${approveResult.errors.map((e) => e.message).join(", ")}`,
        );
        continue;
      }

      currentResource = approveResult.data!.executeActivity;

      // Publish content
      logInfo("Publishing content...");
      const publishResult = await client.executeActivity({
        resourceId: currentResource.id,
        activityId: "publish_content",
        data: {
          action: "publish_content",
          timestamp: new Date().toISOString(),
          publishUrl: `https://blog.example.com/${currentResource.data.title.toLowerCase().replace(/\s+/g, "-")}`,
          seoOptimized: true,
        },
      });

      if (publishResult.errors) {
        logError(
          `Failed to publish content: ${publishResult.errors.map((e) => e.message).join(", ")}`,
        );
        continue;
      }

      currentResource = publishResult.data!.executeActivity;
      logSuccess(`Content published: ${currentResource.data.title}`);

      // Show resource history
      logInfo("Resource lifecycle history:");
      currentResource.history.forEach((event, histIndex) => {
        const timestamp = new Date(event.timestamp).toLocaleTimeString();
        console.log(
          `  ${histIndex + 1}. ${event.fromState} → ${event.toState} via ${event.activity} (${timestamp})`,
        );
      });
    }

    console.log();
    logInfo("Resource Demo Summary:");
    console.log("  • Created multiple content resources with different data");
    console.log("  • Demonstrated complex workflow with revision cycles");
    console.log("  • Showed resource state transitions and data updates");
    console.log("  • Complete audit trail for each resource lifecycle");
    console.log("  • TypeScript integration with GraphQL API");
  } catch (error) {
    logError(`Demo failed: ${error}`);
    process.exit(1);
  }
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch(console.error);
}

export { main as resourceDemo };

File: examples/rust/token_demo.rs
// Resource operations demonstration - Rust GraphQL Client
// Shows detailed resource lifecycle operations using GraphQL API
// Run with: cargo run --example token_demo

use serde_json::json;

// GraphQL endpoint
const GRAPHQL_ENDPOINT: &str = "http://localhost:4000/graphql";

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("🎯 Circuit Breaker Resource Operations Demo - Rust GraphQL Client");
    println!("================================================================");
    println!("Demonstrating: Rust → GraphQL → NATS Backend");
    println!("(Same API as TypeScript, Python, etc.)");
    println!();

    // Create HTTP client
    let client = reqwest::Client::new();

    // 1. Create e-commerce order fulfillment workflow
    println!("📋 Creating E-commerce Order Fulfillment Workflow...");
    let workflow_definition = json!({
        "name": "E-commerce Order Fulfillment",
        "states": [
            "cart", "payment_pending", "paid", "shipped", "delivered", "cancelled"
        ],
        "activities": [
            {
                "id": "checkout",
                "fromStates": ["cart"],
                "toState": "payment_pending",
                "conditions": []
            },
            {
                "id": "pay",
                "fromStates": ["payment_pending"],
                "toState": "paid",
                "conditions": []
            },
            {
                "id": "ship",
                "fromStates": ["paid"],
                "toState": "shipped",
                "conditions": []
            },
            {
                "id": "deliver",
                "fromStates": ["shipped"],
                "toState": "delivered",
                "conditions": []
            },
            {
                "id": "cancel",
                "fromStates": ["cart", "payment_pending"],
                "toState": "cancelled",
                "conditions": []
            }
        ],
        "initialState": "cart"
    });

    let create_workflow_query = json!({
        "query": r#"
            mutation CreateWorkflow($input: WorkflowDefinitionInput!) {
                createWorkflow(input: $input) {
                    id
                    name
                    states
                    activities {
                        id
                        fromStates
                        toState
                    }
                    initialState
                }
            }
        "#,
        "variables": {
            "input": workflow_definition
        }
    });

    let workflow_result: serde_json::Value = client
        .post(GRAPHQL_ENDPOINT)
        .json(&create_workflow_query)
        .send()
        .await?
        .json()
        .await?;

    if let Some(errors) = workflow_result.get("errors") {
        println!("❌ Failed to create workflow: {}", errors);
        return Ok(());
    }

    let workflow = &workflow_result["data"]["createWorkflow"];
    let workflow_id = workflow["id"].as_str().unwrap();
    println!(
        "✅ Created workflow: {} ({})",
        workflow["name"], workflow_id
    );
    println!();

    // 2. Create multiple order resources
    println!("📦 Creating order resources...");
    let orders = vec![
        json!({
            "order_id": "ORD-12345",
            "customer_id": "CUST-789",
            "customer_name": "Alice Johnson",
            "items": [
                {"product": "Laptop", "price": 999.99, "quantity": 1},
                {"product": "Mouse", "price": 29.99, "quantity": 2}
            ],
            "total": 1059.97,
            "payment_method": "credit_card"
        }),
        json!({
            "order_id": "ORD-12346",
            "customer_id": "CUST-790",
            "customer_name": "Bob Smith",
            "items": [
                {"product": "Keyboard", "price": 149.99, "quantity": 1}
            ],
            "total": 149.99,
            "payment_method": "paypal"
        }),
        json!({
            "order_id": "ORD-12347",
            "customer_id": "CUST-791",
            "customer_name": "Carol Davis",
            "items": [
                {"product": "Monitor", "price": 299.99, "quantity": 2},
                {"product": "Cables", "price": 19.99, "quantity": 3}
            ],
            "total": 659.95,
            "payment_method": "bank_transfer"
        }),
    ];

    let mut resource_ids = Vec::new();

    for (i, order_data) in orders.iter().enumerate() {
        let create_resource_query = json!({
            "query": r#"
                mutation CreateResource($input: ResourceCreateInput!) {
                    createResource(input: $input) {
                        id
                        workflowId
                        state
                        data
                        metadata
                        createdAt
                    }
                }
            "#,
            "variables": {
                "input": {
                    "workflowId": workflow_id,
                    "initialState": "cart",
                    "data": order_data,
                    "metadata": {
                        "created_by": "rust-demo",
                        "priority": if i == 0 { "high" } else { "normal" },
                        "source": "website"
                    }
                }
            }
        });

        let resource_result: serde_json::Value = client
            .post(GRAPHQL_ENDPOINT)
            .json(&create_resource_query)
            .send()
            .await?
            .json()
            .await?;

        if let Some(errors) = resource_result.get("errors") {
            println!("❌ Failed to create resource: {}", errors);
            continue;
        }

        let resource = &resource_result["data"]["createResource"];
        let resource_id = resource["id"].as_str().unwrap();
        resource_ids.push(resource_id.to_string());

        println!(
            "✅ Created order: {} ({})",
            resource["data"]["order_id"], resource_id
        );
    }

    println!("📊 Created {} order resources", resource_ids.len());
    println!();

    // 3. Process orders through different lifecycle paths
    for (i, resource_id) in resource_ids.iter().enumerate() {
        println!("⚡ Processing order {} ({})...", i + 1, resource_id);

        // Get current resource to show order details
        let get_resource_query = json!({
            "query": r#"
                query GetResource($id: String!) {
                    resource(id: $id) {
                        id
                        state
                        data
                    }
                }
            "#,
            "variables": {
                "id": resource_id
            }
        });

        let resource_result: serde_json::Value = client
            .post(GRAPHQL_ENDPOINT)
            .json(&get_resource_query)
            .send()
            .await?
            .json()
            .await?;

        if let Some(resource) = resource_result["data"]["resource"].as_object() {
            println!("   📋 Order ID: {}", resource["data"]["order_id"]);
            println!("   👤 Customer: {}", resource["data"]["customer_name"]);
            println!("   💰 Total: ${}", resource["data"]["total"]);
            println!("   📍 Current state: {}", resource["state"]);
        }

        // Determine order path based on index
        let activities = if i == 2 {
            // Third order gets cancelled
            vec![("cancel", "Cancel Order")]
        } else {
            // Normal fulfillment path
            vec![
                ("checkout", "Proceed to Checkout"),
                ("pay", "Process Payment"),
                ("ship", "Ship Order"),
                ("deliver", "Deliver Order"),
            ]
        };

        // Execute activities
        for (activity_id, description) in activities {
            println!("   ➡️  {}", description);

            let execute_activity_query = json!({
                "query": r#"
                    mutation ExecuteActivity($input: ActivityExecuteInput!) {
                        executeActivity(input: $input) {
                            id
                            state
                            data
                            history {
                                timestamp
                                activity
                                fromState
                                toState
                            }
                        }
                    }
                "#,
                "variables": {
                    "input": {
                        "resourceId": resource_id,
                        "activityId": activity_id,
                        "data": {
                            "timestamp": chrono::Utc::now().to_rfc3339(),
                            "processed_by": "rust-demo",
                            "activity": activity_id
                        }
                    }
                }
            });

            match client
                .post(GRAPHQL_ENDPOINT)
                .json(&execute_activity_query)
                .send()
                .await?
                .json::<serde_json::Value>()
                .await?
            {
                activity_result => {
                    if let Some(data) = activity_result.get("data") {
                        let updated_resource = &data["executeActivity"];
                        println!("   ✅ New state: {}", updated_resource["state"]);

                        // Simulate processing time
                        tokio::time::sleep(tokio::time::Duration::from_millis(300)).await;
                    } else if let Some(errors) = activity_result.get("errors") {
                        println!("   ❌ Activity failed: {}", errors[0]["message"]);
                        break;
                    }
                }
            }
        }

        println!();
    }

    // 4. Query final states and show summary
    println!("📊 Final Order Summary:");
    println!("=======================");

    for (i, resource_id) in resource_ids.iter().enumerate() {
        let get_final_resource_query = json!({
            "query": r#"
                query GetResource($id: String!) {
                    resource(id: $id) {
                        id
                        state
                        data
                        history {
                            timestamp
                            activity
                            fromState
                            toState
                        }
                    }
                }
            "#,
            "variables": {
                "id": resource_id
            }
        });

        let final_result: serde_json::Value = client
            .post(GRAPHQL_ENDPOINT)
            .json(&get_final_resource_query)
            .send()
            .await?
            .json()
            .await?;

        if let Some(resource) = final_result["data"]["resource"].as_object() {
            println!("Order {} - {}:", i + 1, resource["data"]["order_id"]);
            println!("   Customer: {}", resource["data"]["customer_name"]);
            println!("   Final State: {}", resource["state"]);
            println!("   Total: ${}", resource["data"]["total"]);

            if let Some(history) = resource["history"].as_array() {
                println!("   Lifecycle ({} steps):", history.len());
                for (j, event) in history.iter().enumerate() {
                    let timestamp = event["timestamp"].as_str().unwrap_or("");
                    let time_part = if let Some(t_pos) = timestamp.find('T') {
                        &timestamp[t_pos + 1..t_pos + 9]
                    } else {
                        timestamp
                    };

                    println!(
                        "     {}. {} → {} via {} ({})",
                        j + 1,
                        event["fromState"],
                        event["toState"],
                        event["activity"],
                        time_part
                    );
                }
            }
            println!();
        }
    }

    println!("🎯 Resource Operations Demo demonstrates:");
    println!("   • Multiple resource creation with different data");
    println!("   • Parallel resource processing");
    println!("   • Different lifecycle paths (normal vs cancelled)");
    println!("   • Complete audit trail for each resource");
    println!("   • GraphQL API consistency across languages");
    println!("   • NATS-backed persistent state management");
    println!();

    println!("💡 This shows the same patterns as:");
    println!("   • TypeScript: npm run demo:token");
    println!("   • Python: python examples/python/token_demo.py");
    println!("   • Any language with GraphQL support!");

    Ok(())
}
